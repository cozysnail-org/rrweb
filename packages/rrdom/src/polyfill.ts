import { RRDocument } from './document';

/**
 * Polyfill the performance for nodejs.
 */
export function polyfillPerformance() {
  if (typeof window !== 'undefined' || 'performance' in global) return;
  ((global as Window & typeof globalThis)
    .performance as unknown) = require('perf_hooks').performance;
}

/**
 * Polyfill requestAnimationFrame and cancelAnimationFrame for nodejs.
 */
export function polyfillRAF() {
  if (typeof window !== 'undefined' || 'requestAnimationFrame' in global)
    return;

  const FPS = 60,
    INTERVAL = 1_000 / FPS;
  let timeoutHandle: NodeJS.Timeout | null = null,
    rafCount = 0,
    requests = Object.create(null);

  function onFrameTimer() {
    const currentRequests = requests;
    requests = Object.create(null);
    timeoutHandle = null;
    Object.keys(currentRequests).forEach(function (id) {
      const request = currentRequests[id];
      if (request) request(Date.now());
    });
  }

  function requestAnimationFrame(callback: (timestamp: number) => void) {
    const cbHandle = ++rafCount;
    requests[cbHandle] = callback;
    if (timeoutHandle === null)
      timeoutHandle = setTimeout(onFrameTimer, INTERVAL);
    return cbHandle;
  }

  function cancelAnimationFrame(handleId: number) {
    delete requests[handleId];
    if (Object.keys(requests).length === 0 && timeoutHandle !== null) {
      clearTimeout(timeoutHandle);
      timeoutHandle = null;
    }
  }

  (global as Window &
    typeof globalThis).requestAnimationFrame = requestAnimationFrame;
  (global as Window &
    typeof globalThis).cancelAnimationFrame = cancelAnimationFrame;
}

/**
 *  Polyfill document object with RRDocument for nodejs.
 */
export function polyfillDocument() {
  if (typeof document !== 'undefined') return;
  const rrdom = new RRDocument();
  (() => {
    rrdom.appendChild(rrdom.createElement('html'));
    rrdom.documentElement.appendChild(rrdom.createElement('head'));
    rrdom.documentElement.appendChild(rrdom.createElement('body'));
  })();
  global.document = (rrdom as unknown) as Document;
}
