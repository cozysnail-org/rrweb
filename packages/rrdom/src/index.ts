import { polyfillPerformance, polyfillRAF, polyfillDocument } from './polyfill';
polyfillPerformance();
polyfillRAF();
polyfillDocument();
export * from './document';
