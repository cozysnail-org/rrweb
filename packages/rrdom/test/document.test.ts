/**
 * @jest-environment jsdom
 */
import * as fs from 'fs';
import * as path from 'path';
import { RRDocument, RRElement } from '../src/document';
import { printRRDom } from './util';

describe('RRDocument', () => {
  describe('buildFromDom', () => {
    it('should create an RRDocument from a html document', () => {
      // setup document
      document.write(getHtml('main.html'));

      // create RRDocument from document
      const rrdoc = new RRDocument();
      rrdoc.buildFromDom(document);
      expect(printRRDom(rrdoc)).toMatchSnapshot();
    });
  });
});

function getHtml(fileName: string) {
  const filePath = path.resolve(__dirname, `./html/${fileName}`);
  return fs.readFileSync(filePath, 'utf8');
}
