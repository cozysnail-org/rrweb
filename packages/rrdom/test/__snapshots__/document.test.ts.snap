// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RRDocument buildFromDom should create an RRDocument from a html document 1`] = `
"-1 RRDocument
  -2 RRDocumentType
  -3 HTML lang=\\"en\\" 
    -4 HEAD 
      -5 RRText text=\\"\\\\n    \\"
      -6 META charset=\\"UTF-8\\" 
      -7 RRText text=\\"\\\\n    \\"
      -8 META name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0\\" 
      -9 RRText text=\\"\\\\n    \\"
      -10 TITLE 
        -11 RRText text=\\"Main\\"
      -12 RRText text=\\"\\\\n    \\"
      -13 STYLE 
        -14 RRText text=\\"\\\\n      h1...\\"
      -15 RRText text=\\"\\\\n  \\"
    -16 RRText text=\\"\\\\n  \\"
    -17 BODY 
      -18 RRText text=\\"\\\\n    \\"
      -19 H1 
        -20 RRText text=\\"This is a ...\\"
      -21 RRText text=\\"\\\\n    \\"
      -22 H1 style=\\"font-size: 16px\\" 
        -23 RRText text=\\"This is a ...\\"
      -24 RRText text=\\"\\\\n    \\"
      -25 DIV id=\\"block1\\" class=\\"blocks blocks1\\" 
        -26 RRText text=\\"\\\\n      \\"
        -27 DIV id=\\"block2\\" class=\\"blocks blocks1\\" 
          -28 RRText text=\\"\\\\n        ...\\"
          -29 DIV id=\\"block3\\" 
            -30 RRText text=\\"\\\\n        ...\\"
            -31 P 
              -32 RRText text=\\"This is a ...\\"
            -33 RRText text=\\"\\\\n        ...\\"
            -34 BUTTON 
              -35 RRText text=\\"button1\\"
            -36 RRText text=\\"\\\\n        \\"
          -37 RRText text=\\"\\\\n        ...\\"
        -38 RRText text=\\"\\\\n      \\"
        -39 IMG src=\\"somelink\\" alt=\\"This is an image\\" 
        -40 RRText text=\\"\\\\n    \\"
      -41 RRText text=\\"\\\\n  

\\"
"
`;
